experimental: Add health server support

joamaki opened the pull request:
This implements the HealthCheckNodePort functionality for running a HTTP health server that responds with the number of active backends a service has available.

joamaki committed:
experimental: Add health server support

This implements the HealthCheckNodePort functionality for running
a HTTP health server that responds with the number of active backends
a service has available.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

ovidiutirla commented on pkg/loadbalancer/experimental/config.go:
```diff
@@ -40,30 +41,51 @@ type TestConfig struct {
 func (def TestConfig) Flags(flags *pflag.FlagSet) {
 	flags.Float32("lb-test-fault-probability", def.TestFaultProbability, "Probability for fault injection in LBMaps")
 	flags.String("node-port-algorithm", option.NodePortAlgRandom, "NodePort algorithm")
+	flags.Bool("enable-health-check-nodeport", false, "Enable the NodePort health check server")
```
mark hidden?

joamaki replied:
The TestConfig is only used by tests so no need to hide it.

ovidiutirla reviewed (approved):
🚀 nice improvement

julianwiedmann commented on pkg/loadbalancer/experimental/healthserver.go:
```diff
@@ -0,0 +1,215 @@
+// For each service that has HealthCheckNodePort set to a non-zero value it
+// starts an HTTP server on that port and responds with the number of active
+// backends. This allows external load-balancers to avoid directing traffic
+// to services that have no available backends.
```
From my understanding of the feature:

```suggestion
// For each service that has HealthCheckNodePort set to a non-zero value it
// starts an HTTP server on that port and responds with the number of active
// backends on the node. This allows external load-balancers to avoid directing traffic
// to nodes that have no available backends.
```

julianwiedmann commented on pkg/loadbalancer/experimental/healthserver.go:
```diff
@@ -0,0 +1,215 @@
+				}
+				s.serverByPort[port] = s.addListener(name, port)
+				s.portByService[name] = port
+			}
```
I got hung up on this a bit, that `found = false` part is difficult :slightly_smiling_face:.
I broke it down for myself into what's below to confirm the logic, maybe you want to pick it up?

```suggestion
				oldPort, found := s.portByService[name]
			       // HealthCheckNodePort has changed, we treat this as a DeleteService
			       // followed by an Insert.

				if found && oldPort == port {
				        // no change needed
				        continue
				}

                               // delete old port
			       if (found) {
                                        s.removeListener(ctx, oldPort)
					delete(s.portByService, name)
				}

                               // add new port
				if (port) {
				        s.serverByPort[port] = s.addListener(name, port)
				        s.portByService[name] = port
			       }
```

joamaki replied:
Thanks, reworked it!

julianwiedmann commented on pkg/loadbalancer/experimental/testdata/healthserver.txtar:
```diff
@@ -0,0 +1,240 @@
+Content-Type=application/json
+Date=<omitted>
+X-Content-Type-Options=nosniff
+X-Load-Balancing-Endpoint-Weight=4
```
From my quick read of the expected behavior I'd have assumed this to be

```suggestion
X-Load-Balancing-Endpoint-Weight=2
```

julianwiedmann reviewed (approved):
lgtm overall, assuming the returned count of backends (total vs local) is the actual expected behavior. Otherwise that's an easy follow-up fix :).

julianwiedmann commented on pkg/loadbalancer/experimental/healthserver.go:
```diff
@@ -0,0 +1,296 @@
+		}
+		if be.State == lb.BackendStateActive {
+			activeCount++
+		}
```
looking at the old `upsertService()` code - did we lose the special-casing for L7 services here? It's always reporting 1 active backend (representing the envoy instance).

joamaki replied:
added.

julianwiedmann commented on pkg/loadbalancer/experimental/healthserver.go:
```diff
@@ -0,0 +1,296 @@
+			if !isExtLocal(fe) {
+				continue
+			}
+
```
Old `upsertService()` code seems to support that a service was marked `eTP:local` in the past, but then gets modified. I suppose that's a bit too crazy to continue supporting?

joamaki replied:
Fixed it to support this.

julianwiedmann commented on pkg/loadbalancer/experimental/healthserver.go:
```diff
@@ -0,0 +1,296 @@
+						Scope: lb.ScopeExternal,
+					},
+					Type:        fe.Type,
+					ServiceName: name,
```
Should this be

```suggestion
					ServiceName: healthServiceName,
```

joamaki replied:
yep, this was fixed.

julianwiedmann reviewed (commented):
Thank you! Some comments inline, primarily how L7-LB is handled.

joamaki committed:
experimental: Filter backends when traffic policy Local

Do not include non-local backends when the traffic policy is local.
Add testdata/trafficpolicy.txtar test case.

To reduce idle-state memory usage switch from []BackendWithRevision to
iter.Seq2[*Backend, Revision].

Signed-off-by: Jussi Maki <jussi@isovalent.com>

sayboras reviewed (approved):
Looks good for envoy related file

brb commented on pkg/loadbalancer/experimental/writer.go:
```diff
@@ -582,3 +594,43 @@ func (w *Writer) DebugDump(txn statedb.ReadTxn, to io.Writer) {
+}
+
+func isIntLocal(fe *Frontend) bool {
+	/* FIXME if !option.Config.EnableInternalTrafficPolicy {
```
OT to your PR - any idea why we have a flag for InternalTrafficPolicy?

joamaki replied:
for feature gating purposes

julianwiedmann reviewed (approved):
:rocket: lgtm!
